internal/replication/blockreplicator_3node_test.go

```diff
@@ -0,0 +1,313 @@
+	}
+	assert.Eventually(t, isLeaderCond1, 30*time.Second, 100*time.Millisecond)
+
+	err = env.nodes[2].Start()
+	require.NoError(t, err)
+	isLeaderCond2 := func() bool {
+		return env.AgreedLeaderIndex() >= 0
+	}
+	assert.Eventually(t, isLeaderCond2, 30*time.Second, 100*time.Millisecond)
+
+	//close all
+	for _, node := range env.nodes {
+		err := node.Close()
+		require.NoError(t, err)
+	}
+}
+
+// Scenario:
+// - Start 3 nodes together, wait for leader,
+// - Stop the leader, wait for new leader,
+// - Restart the node, wait for consistent leader on all
+// - Stop a follower, wait for new leader,
+// - Restart the node, wait for consistent leader on all
+func TestBlockReplicator_3Node_Restart(t *testing.T) {
+	env := createClusterEnv(t, "info", 3)
+	require.Equal(t, 3, len(env.nodes))
+	defer os.RemoveAll(env.testDir)
+
+	// start 3 at once
+	for _, node := range env.nodes {
+		err := node.Start()
+		require.NoError(t, err)
+	}
+
+	// wait for an agreed leader
+	sameLeaderCond := func() bool {
+		return env.AgreedLeaderIndex() >= 0
+	}
+	assert.Eventually(t, sameLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	//close the leader, wait for some node to become a new leader
+	leaderIndex1 := env.FindLeaderIndex()
+	require.True(t, leaderIndex1 >= 0)
+	t.Logf("Stopping leader, index: %d", leaderIndex1)
+	err := env.nodes[leaderIndex1].Close()
+	require.NoError(t, err)
+	isLeaderCond := func() bool {
+		return env.FindLeaderIndex() >= 0
+	}
+	assert.Eventually(t, isLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	//restart the old leader
+	t.Logf("Restarting node, index: %d", leaderIndex1)
+	err = env.nodes[leaderIndex1].Restart()
+	require.NoError(t, err)
+	assert.Eventually(t, sameLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	leaderIndex2 := env.FindLeaderIndex()
+	require.True(t, leaderIndex2 >= 0)
+	followerIndex := (leaderIndex1 + 1) % 3
+	t.Logf("Stopping follower, index: %d", followerIndex)
+	err = env.nodes[followerIndex].Close()
+	require.NoError(t, err)
+	assert.Eventually(t, isLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	//restart the follower
+	t.Logf("Restarting node, index: %d", followerIndex)
+	err = env.nodes[followerIndex].Restart()
+	require.NoError(t, err)
+	assert.Eventually(t, sameLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	//close all
+	for _, node := range env.nodes {
+		err := node.Close()
+		require.NoError(t, err)
+	}
+}
+
+// Scenario:
+// - Start 3 nodes together, wait for leader,
+// - Submit 100 blocks, wait for all ledgers to get them.
+func TestBlockReplicator_3Node_Submit(t *testing.T) {
+	env := createClusterEnv(t, "info", 3)
+	require.Equal(t, 3, len(env.nodes))
+	defer os.RemoveAll(env.testDir)
+
+	for _, node := range env.nodes {
+		err := node.Start()
+		require.NoError(t, err)
+	}
+
+	// wait for some node to become a leader
+	isLeaderCond := func() bool {
+		return env.FindLeaderIndex() >= 0
+	}
+	assert.Eventually(t, isLeaderCond, 30*time.Second, 100*time.Millisecond)
```

Comment:
as it is repetitive, we can make it into a function and reuse.

Reply:
done